File: tests/p2p_cartridge_properties_test.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath)
//
// BoJ Server — P2P (Point-to-Point) Property Tests
//
// Property-based tests that validate invariants across all cartridges:
// - Every cartridge in the listing has a corresponding info entry
// - All tool signatures conform to JSON Schema
// - Cartridge names are unique
// - Domains and tiers are from the approved vocabulary

import { assertEquals, assertExists } from "https://deno.land/std@0.208.0/assert/mod.ts";

// ===================================================================
// Cartridge Vocabulary (Approved Set)
// ===================================================================

const APPROVED_DOMAINS = [
  "infrastructure",
  "data",
  "orchestration",
  "security",
  "messaging",
  "observability",
  "development",
  "ai",
  "integration",
];

const APPROVED_TIERS = [
  "teranga",   // Hospitality tier (foundational)
  "shield",    // Security tier
  "umoja",     // Unity tier (advanced features)
];

const APPROVED_PROTOCOLS = [
  "json-rpc",
  "rest",
  "grpc",
  "graphql",
  "websocket",
];

// ===================================================================
// Mock Cartridge Catalogue
// ===================================================================

const MOCK_CARTRIDGES = [
  { name: "boj_health", domain: "infrastructure", protocol: "json-rpc", tier: "teranga" },
  { name: "boj_cartridges", domain: "infrastructure", protocol: "json-rpc", tier: "teranga" },
  { name: "database-mcp", domain: "data", protocol: "json-rpc", tier: "teranga" },
  { name: "fleet-mcp", domain: "orchestration", protocol: "json-rpc", tier: "teranga" },
  { name: "nesy-mcp", domain: "ai", protocol: "json-rpc", tier: "teranga" },
  { name: "agent-mcp", domain: "ai", protocol: "json-rpc", tier: "teranga" },
  { name: "cloud-mcp", domain: "orchestration", protocol: "rest", tier: "teranga" },
  { name: "container-mcp", domain: "orchestration", protocol: "json-rpc", tier: "teranga" },
  { name: "k8s-mcp", domain: "orchestration", protocol: "grpc", tier: "umoja" },
  { name: "git-mcp", domain: "development", protocol: "json-rpc", tier: "teranga" },
  { name: "secrets-mcp", domain: "security", protocol: "json-rpc", tier: "shield" },
  { name: "queues-mcp", domain: "messaging", protocol: "json-rpc", tier: "teranga" },
  { name: "iac-mcp", domain: "orchestration", protocol: "json-rpc", tier: "umoja" },
  { name: "observe-mcp", domain: "observability", protocol: "json-rpc", tier: "teranga" },
  { name: "ssg-mcp", domain: "development", protocol: "json-rpc", tier: "teranga" },
  { name: "proof-mcp", domain: "ai", protocol: "json-rpc", tier: "shield" },
  { name: "lsp-mcp", domain: "development", protocol: "json-rpc", tier: "teranga" },
];

// ===================================================================
// P2P: Cartridge Name Uniqueness
// ===================================================================

Deno.test("P2P: All cartridge names are unique", {
  ignore: false,
}, () => {
  const names = MOCK_CARTRIDGES.map((c) => c.name);
  const uniqueNames = new Set(names);

  assertEquals(
    uniqueNames.size,
    names.length,
    `Duplicate cartridge names found: ${names.join(", ")}`
  );
});

// ===================================================================
// P2P: Domain Vocabulary Compliance
// ===================================================================

Deno.test("P2P: All cartridges use approved domains", {
  ignore: false,
}, () => {
  MOCK_CARTRIDGES.forEach((cartridge) => {
    assertEquals(
      APPROVED_DOMAINS.includes(cartridge.domain),
      true,
      `Unknown domain '${cartridge.domain}' for cartridge '${cartridge.name}'`
    );
  });
});

// ===================================================================
// P2P: Tier Vocabulary Compliance
// ===================================================================

Deno.test("P2P: All cartridges use approved tiers", {
  ignore: false,
}, () => {
  MOCK_CARTRIDGES.forEach((cartridge) => {
    assertEquals(
      APPROVED_TIERS.includes(cartridge.tier),
      true,
      `Unknown tier '${cartridge.tier}' for cartridge '${cartridge.name}'`
    );
  });
});

// ===================================================================
// P2P: Protocol Vocabulary Compliance
// ===================================================================

Deno.test("P2P: All cartridges use approved protocols", {
  ignore: false,
}, () => {
  MOCK_CARTRIDGES.forEach((cartridge) => {
    assertEquals(
      APPROVED_PROTOCOLS.includes(cartridge.protocol),
      true,
      `Unknown protocol '${cartridge.protocol}' for cartridge '${cartridge.name}'`
    );
  });
});

// ===================================================================
// P2P: Tool Schema Compliance
// ===================================================================

Deno.test("P2P: All tools have required schema fields", {
  ignore: false,
}, () => {
  const tools = [
    {
      name: "db_query",
      description: "Query a database",
      inputSchema: {
        type: "object",
        properties: {
          sql: { type: "string" },
        },
        required: ["sql"],
      },
    },
    {
      name: "fleet_status",
      description: "Get fleet status",
      inputSchema: {
        type: "object",
        properties: {
          fleet_id: { type: "string" },
        },
        required: ["fleet_id"],
      },
    },
  ];

  tools.forEach((tool) => {
    assertEquals(typeof tool.name, "string");
    assertEquals(typeof tool.description, "string");
    assertExists(tool.inputSchema);
    assertEquals(tool.inputSchema.type, "object");
    assertExists(tool.inputSchema.properties);
    assertExists(tool.inputSchema.required);
    assertEquals(Array.isArray(tool.inputSchema.required), true);
  });
});

// ===================================================================
// P2P: Tool Name Uniqueness Within Cartridge
// ===================================================================

Deno.test("P2P: Tool names are unique within cartridge", {
  ignore: false,
}, () => {
  const cartridgeTools = {
    "database-mcp": [
      { name: "db_query" },
      { name: "db_connect" },
      { name: "db_disconnect" },
    ],
  };

  Object.entries(cartridgeTools).forEach(([cartridgeName, tools]) => {
    const names = tools.map((t) => t.name);
    const uniqueNames = new Set(names);

    assertEquals(
      uniqueNames.size,
      names.length,
      `Duplicate tool names in ${cartridgeName}: ${names.join(", ")}`
    );
  });
});

// ===================================================================
// P2P: Input Schema Property Types
// ===================================================================

Deno.test("P2P: All inputSchema properties have valid types", {
  ignore: false,
}, () => {
  const validTypes = ["string", "number", "boolean", "object", "array"];

  const testSchema = {
    type: "object",
    properties: {
      query: { type: "string" },
      limit: { type: "number" },
      active: { type: "boolean" },
      filters: { type: "object" },
      tags: { type: "array" },
    },
  };

  Object.entries(testSchema.properties).forEach(([prop, spec]) => {
    assertEquals(
      validTypes.includes(spec.type),
      true,
      `Invalid type '${spec.type}' for property '${prop}'`
    );
  });
});

// ===================================================================
// P2P: Cartridge to Domain Mapping
// ===================================================================

Deno.test("P2P: Domain presence property", {
  ignore: false,
}, () => {
  // Each domain should have at least one cartridge
  const domainMap = new Map<string, number>();

  MOCK_CARTRIDGES.forEach((c) => {
    domainMap.set(c.domain, (domainMap.get(c.domain) ?? 0) + 1);
  });

  assertEquals(domainMap.size > 0, true);
  domainMap.forEach((count) => {
    assertEquals(count > 0, true);
  });
});

// ===================================================================
// P2P: Tier Distribution Property
// ===================================================================

Deno.test("P2P: Each tier has at least one cartridge", {
  ignore: false,
}, () => {
  const tierMap = new Map<string, number>();

  MOCK_CARTRIDGES.forEach((c) => {
    tierMap.set(c.tier, (tierMap.get(c.tier) ?? 0) + 1);
  });

  // teranga and shield should exist, umoja may not
  assertEquals(tierMap.has("teranga"), true);
  assertEquals(tierMap.has("shield"), true);
});

// ===================================================================
// P2P: Cartridge Name Format
// ===================================================================

Deno.test("P2P: Cartridge names match naming convention", {
  ignore: false,
}, () => {
  // Names should be lowercase, alphanumeric + dash/underscore
  const namePattern = /^[a-z0-9_-]+$/;

  MOCK_CARTRIDGES.forEach((c) => {
    assertEquals(
      namePattern.test(c.name),
      true,
      `Invalid name format: '${c.name}'`
    );
  });
});

// ===================================================================
// P2P: Cartridge Count Property
// ===================================================================

Deno.test("P2P: Cartridge count is reasonable", {
  ignore: false,
}, () => {
  // Should have at least 2 (health + list), probably 90+
  assertEquals(MOCK_CARTRIDGES.length >= 2, true);
  assertEquals(MOCK_CARTRIDGES.length <= 200, true);
});

// ===================================================================
// P2P: Tool Count Per Cartridge
// ===================================================================

Deno.test("P2P: Tool count is reasonable per cartridge", {
  ignore: false,
}, () => {
  // Each cartridge should have 1-50 tools
  const toolCounts = {
    "database-mcp": 15,
    "fleet-mcp": 12,
    "secrets-mcp": 8,
  };

  Object.entries(toolCounts).forEach(([cartridge, toolCount]) => {
    assertEquals(toolCount >= 1, true);
    assertEquals(toolCount <= 50, true);
  });
});

// ===================================================================
// P2P: Matrix Completeness
// ===================================================================

Deno.test("P2P: Cartridge matrix is well-distributed", {
  ignore: false,
}, () => {
  const matrix = new Map<string, Set<string>>();

  MOCK_CARTRIDGES.forEach((c) => {
    const key = `${c.domain}:${c.protocol}`;
    if (!matrix.has(key)) {
      matrix.set(key, new Set());
    }
    matrix.get(key)!.add(c.name);
  });

  // Matrix should have some coverage
  assertEquals(matrix.size > 0, true);
});

// ===================================================================
// P2P: Cartridge Health Property
// ===================================================================

Deno.test("P2P: Critical cartridges are present", {
  ignore: false,
}, () => {
  const criticalNames = ["boj_health", "boj_cartridges"];
  const names = new Set(MOCK_CARTRIDGES.map((c) => c.name));

  criticalNames.forEach((critical) => {
    assertEquals(
      names.has(critical),
      true,
      `Critical cartridge '${critical}' not found`
    );
  });
});

File: tests/p2p_cartridge_properties_test.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// BoJ Server — P2P Cartridge Property Tests (Zig)
//
// Validates vocabulary invariants across all catalogued cartridges:
// - Domain, tier, and protocol from approved sets
// - Name uniqueness and naming convention
// - Tool schema compliance
// - Matrix coverage
//
// Run: zig test tests/p2p_cartridge_properties_test.zig

const std = @import("std");
const testing = std.testing;

const Cartridge = struct {
    name: []const u8,
    domain: []const u8,
    protocol: []const u8,
    tier: []const u8,
};

const CATALOGUE = [_]Cartridge{
    .{ .name = "boj_health",     .domain = "infrastructure", .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "boj_cartridges", .domain = "infrastructure", .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "database-mcp",   .domain = "data",           .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "fleet-mcp",      .domain = "orchestration",  .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "nesy-mcp",       .domain = "ai",             .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "agent-mcp",      .domain = "ai",             .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "cloud-mcp",      .domain = "orchestration",  .protocol = "rest",     .tier = "teranga" },
    .{ .name = "container-mcp",  .domain = "orchestration",  .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "k8s-mcp",        .domain = "orchestration",  .protocol = "grpc",     .tier = "umoja"   },
    .{ .name = "git-mcp",        .domain = "development",    .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "secrets-mcp",    .domain = "security",       .protocol = "json-rpc", .tier = "shield"  },
    .{ .name = "queues-mcp",     .domain = "messaging",      .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "iac-mcp",        .domain = "orchestration",  .protocol = "json-rpc", .tier = "umoja"   },
    .{ .name = "observe-mcp",    .domain = "observability",  .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "ssg-mcp",        .domain = "development",    .protocol = "json-rpc", .tier = "teranga" },
    .{ .name = "proof-mcp",      .domain = "ai",             .protocol = "json-rpc", .tier = "shield"  },
    .{ .name = "lsp-mcp",        .domain = "development",    .protocol = "json-rpc", .tier = "teranga" },
};

const APPROVED_DOMAINS = [_][]const u8{
    "infrastructure", "data", "orchestration", "security",
    "messaging", "observability", "development", "ai", "integration",
};

const APPROVED_TIERS = [_][]const u8{ "teranga", "shield", "umoja" };

const APPROVED_PROTOCOLS = [_][]const u8{ "json-rpc", "rest", "grpc", "graphql", "websocket" };

fn inSlice(haystack: []const []const u8, needle: []const u8) bool {
    for (haystack) |s| if (std.mem.eql(u8, s, needle)) return true;
    return false;
}

fn isValidName(name: []const u8) bool {
    if (name.len == 0) return false;
    for (name) |c| {
        if (!std.ascii.isAlphanumeric(c) and c != '_' and c != '-') return false;
    }
    return true;
}

test "P2P: all cartridge names are unique" {
    for (CATALOGUE, 0..) |a, i| {
        for (CATALOGUE, 0..) |b, j| {
            if (i == j) continue;
            try testing.expect(!std.mem.eql(u8, a.name, b.name));
        }
    }
}

test "P2P: all cartridges use approved domains" {
    for (CATALOGUE) |c| {
        const ok = inSlice(&APPROVED_DOMAINS, c.domain);
        if (!ok) std.debug.print("bad domain: {s}\n", .{c.domain});
        try testing.expect(ok);
    }
}

test "P2P: all cartridges use approved tiers" {
    for (CATALOGUE) |c| {
        const ok = inSlice(&APPROVED_TIERS, c.tier);
        if (!ok) std.debug.print("bad tier: {s} for {s}\n", .{ c.tier, c.name });
        try testing.expect(ok);
    }
}

test "P2P: all cartridges use approved protocols" {
    for (CATALOGUE) |c| {
        const ok = inSlice(&APPROVED_PROTOCOLS, c.protocol);
        if (!ok) std.debug.print("bad protocol: {s}\n", .{c.protocol});
        try testing.expect(ok);
    }
}

test "P2P: all cartridge names match naming convention" {
    for (CATALOGUE) |c| {
        try testing.expect(isValidName(c.name));
    }
}

test "P2P: tool schema required fields are present" {
    const Tool = struct {
        name: []const u8,
        description: []const u8,
        input_type: []const u8, // must be "object"
    };
    const tools = [_]Tool{
        .{ .name = "db_query",    .description = "Query a database", .input_type = "object" },
        .{ .name = "fleet_status",.description = "Get fleet status", .input_type = "object" },
    };
    for (tools) |t| {
        try testing.expect(t.name.len > 0);
        try testing.expect(t.description.len > 0);
        try testing.expectEqualStrings("object", t.input_type);
    }
}

test "P2P: tool names are unique within a cartridge" {
    const db_tools = [_][]const u8{ "db_query", "db_connect", "db_disconnect" };
    for (db_tools, 0..) |a, i| {
        for (db_tools, 0..) |b, j| {
            if (i == j) continue;
            try testing.expect(!std.mem.eql(u8, a, b));
        }
    }
}

test "P2P: valid JSON schema property types" {
    const valid_types = [_][]const u8{ "string", "number", "boolean", "object", "array" };
    const test_types = [_][]const u8{ "string", "number", "boolean", "object", "array" };
    for (test_types) |t| try testing.expect(inSlice(&valid_types, t));
}

test "P2P: teranga tier has at least one cartridge" {
    var found = false;
    for (CATALOGUE) |c| if (std.mem.eql(u8, c.tier, "teranga")) { found = true; break; };
    try testing.expect(found);
}

test "P2P: shield tier has at least one cartridge" {
    var found = false;
    for (CATALOGUE) |c| if (std.mem.eql(u8, c.tier, "shield")) { found = true; break; };
    try testing.expect(found);
}

test "P2P: critical cartridges boj_health and boj_cartridges are present" {
    const critical = [_][]const u8{ "boj_health", "boj_cartridges" };
    for (critical) |req| {
        var found = false;
        for (CATALOGUE) |c| if (std.mem.eql(u8, c.name, req)) { found = true; break; };
        try testing.expect(found);
    }
}

test "P2P: cartridge count is in reasonable range [2, 200]" {
    try testing.expect(CATALOGUE.len >= 2);
    try testing.expect(CATALOGUE.len <= 200);
}

test "P2P: domain coverage spans multiple domains" {
    var seen_domains: usize = 0;
    for (APPROVED_DOMAINS) |dom| {
        for (CATALOGUE) |c| {
            if (std.mem.eql(u8, c.domain, dom)) { seen_domains += 1; break; }
        }
    }
    try testing.expect(seen_domains >= 3);
}
